pkg/virt-controller/watch/export/export.go

```diff
@@ -328,77 +362,204 @@ func (ctrl *VMExportController) getOwnerVMexportKey(obj metav1.Object) string {
 
 	if vmExport.DeletionTimestamp != nil {
-		return retry, nil
+		return nil
 	}
 
 	service, err := ctrl.getOrCreateExportService(vmExport)
 	if err != nil {
-		return 0, err
+		return err
 	}
 
 	if ctrl.isSourcePvc(&vmExport.Spec) {
 		return ctrl.handleIsSourcePvc(vmExport, service)
+	} else if ctrl.isSourceVMSnapshot(&vmExport.Spec) {
+		return ctrl.handleIsVmSnapshot(vmExport, service)
 	}
-	return retry, nil
+	return nil
 }
 
-func (ctrl *VMExportController) handleIsSourcePvc(vmExport *exportv1.VirtualMachineExport, service *corev1.Service) (time.Duration, error) {
-	var retry time.Duration
+func (ctrl *VMExportController) handleIsSourcePvc(vmExport *exportv1.VirtualMachineExport, service *corev1.Service) error {
 	pvcs := make([]*corev1.PersistentVolumeClaim, 0)
 	pvc, exists, err := ctrl.getPvc(vmExport.Namespace, vmExport.Spec.Source.Name)
 	if err != nil {
-		return 0, err
+		return err
 	} else if exists {
 		pvcs = append(pvcs, pvc)
 	}
 
 	pod, exists, err := ctrl.getExporterPod(vmExport)
 	inUse := false
 	if err != nil {
-		return 0, err
+		return err
 	} else if !exists {
 		inUse, err = ctrl.isPVCInUse(vmExport, pvc)
 		if err != nil {
-			return retry, err
+			return err
 		}
 		if !inUse && len(pvcs) > 0 {
 			isPopulated, err := ctrl.isPVCPopulated(pvc)
 			if err != nil {
-				return retry, err
+				return err
 			} else if isPopulated {
 				pod, err = ctrl.createExporterPod(vmExport, pvcs)
 				if err != nil {
-					return 0, err
+					return err
 				} else if pod == nil {
-					return retry, nil
+					return nil
 				}
 
 				if err := ctrl.getOrCreateCertSecret(vmExport, pod); err != nil {
-					return 0, err
+					return err
 				}
 			}
 		}
 	} else {
-		if pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed {
-			// The server died or completed, delete the pod.
-			ctrl.Recorder.Eventf(vmExport, corev1.EventTypeWarning, exporterPodFailedOrCompletedEvent, "Exporter pod %s/%s succeeded or failed", pod.Namespace, pod.Name)
-			if err := ctrl.Client.CoreV1().Pods(vmExport.Namespace).Delete(context.Background(), pod.Name, metav1.DeleteOptions{}); err != nil {
-				return 0, err
-			} else {
-				return retry, nil
-			}
+		if err := ctrl.handlePodSucceededOrFailed(vmExport, pod); err != nil {
+			return err
 		}
 	}
 
 	return ctrl.updateVMExportPvcStatus(vmExport, pvcs, pod, service, inUse)
 
 }
 
+func (ctrl *VMExportController) handleIsVmSnapshot(vmExport *exportv1.VirtualMachineExport, service *corev1.Service) error {
+	var pvcs []*corev1.PersistentVolumeClaim
+
+	vmSnapshot, exists, err := ctrl.getVmSnapshot(vmExport.Namespace, vmExport.Spec.Source.Name)
+	restoreableSnapshots := 0
+	if err != nil {
+		return err
+	} else if exists {
+		if vmSnapshot.Status.ReadyToUse != nil && *vmSnapshot.Status.ReadyToUse {
+			pvcs, restoreableSnapshots, err = ctrl.handlePVCsForVirtualMachineSnapshot(vmExport, vmSnapshot)
+			if err != nil {
+				return err
+			}
+		} else {
+			return nil
```

Comment:
Should status be updated?

Reply:
Good point, I can just fall through the same way as if the snapshot doesn't exist. The result is the same.